# 1.0 Data Cleaning Walkthrough

In [0]:
from google.colab import files
uploaded = files.upload()
for fn in uploaded.keys():
  print('User uploaded file "{name}" with length {length} bytes'.format(
      name=fn, length=len(uploaded[fn])))

Saving ap_2010.csv to ap_2010.csv
Saving class_size.csv to class_size.csv
Saving demographics.csv to demographics.csv
Saving graduation.csv to graduation.csv
Saving hs_directory.csv to hs_directory.csv
Saving sat_results.csv to sat_results.csv
Saving survey_all.txt to survey_all.txt
Saving survey_d75.txt to survey_d75.txt
User uploaded file "ap_2010.csv" with length 11734 bytes
User uploaded file "class_size.csv" with length 2501522 bytes
User uploaded file "demographics.csv" with length 1578546 bytes
User uploaded file "graduation.csv" with length 3145877 bytes
User uploaded file "hs_directory.csv" with length 1342523 bytes
User uploaded file "sat_results.csv" with length 28339 bytes
User uploaded file "survey_all.txt" with length 8431567 bytes
User uploaded file "survey_d75.txt" with length 292057 bytes


In [0]:
from google.colab import drive
drive.mount('/content/drive')


Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope=email%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdocs.test%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive.photos.readonly%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fpeopleapi.readonly&response_type=code

Enter your authorization code:
··········
Mounted at /content/drive


## 1.1 Introduction

At many points in your career, you'll need to be able to build complete, end-to-end data science projects on your own. **Data science projects usually consist of one of two things:**

- An exploration and analysis of a set of data. One example might involve analyzing donors to political campaigns, creating a plot, and then sharing an analysis of the plot with others.
- An operational system that generates predictions based on data that updates continually. An algorithm that pulls in daily stock ticker data and predicts which stock prices will rise and fall would be one example.

You'll find the ability to create data science projects useful in several different contexts:

- Projects will help you build a portfolio, which is critical to finding a job as a data analyst or scientist.
- Working on projects will help you learn new skills and reinforce existing concepts.
- Most "real-world" data science and analysis work consists of developing internal projects.
- Projects allow you to investigate interesting phenomena and satisfy your curiosity.

Whether you aim to become a data scientist or analyst or you're just curious about the world, building projects can be immensely rewarding.

**In this section, we'll walk through the first part of a complete data science project, including how to acquire the raw data.** The project will focus on exploring and analyzing a data set. We'll develop our **data cleaning** and **storytelling skills**, which will enable us to build complete projects on our own.

We'll focus **primarily on data exploration** in this section. We'll also combine several messy data sets into a single clean one to make analysis easier. Over the next few missions, we'll work through the rest of our project and perform the actual analysis.

**The first step in creating a project is to decide on a topic**. You want the topic to be something you're interested in and motivated to explore. It's very obvious when people are making projects just to make them, rather than out of a genuine interest in the topic.

Here are two ways to go about finding a good topic:

- Think about what sectors or angles you're really interested in, then find data sets relating to those sectors.
- Review several data sets, and find one that seems interesting enough to explore.

Whichever approach you take, you can start your search at these sites:

- [Data.gov](https://www.data.gov/) - A directory of government data downloads
- [/r/datasets](https://reddit.com/r/datasets) - A subreddit that has hundreds of interesting data sets
- [Awesome datasets](https://github.com/caesar0301/awesome-public-datasets) - A list of data sets hosted on GitHub
- [rs.io](http://rs.io/100-interesting-data-sets-for-statistics/) - A great blog post with hundreds of interesting data sets
In real-world data science, you may not find an ideal data set. You might have to aggregate disparate data sources instead, or do a good amount of data cleaning.

For the purposes of this project, we'll be using data about New York City public schools, which can be found [here](https://data.cityofnewyork.us/browse?category=Education).

## 1.2 Finding All of the Relevant Data Sets

Once you've chosen a topic, you'll want to pick an angle to investigate. It's important to choose an angle that has enough depth to analyze, but isn't so complicated that it's difficult to get started. You want to finish the project, and you want your results to be interesting to others.

One of the most controversial issues in the U.S. educational system is the efficacy of standardized tests, and whether they're unfair to certain groups. Given our prior knowledge of this topic, investigating the correlations between [SAT scores](https://en.wikipedia.org/wiki/SAT) and demographics might be an interesting angle to take. We could correlate SAT scores with factors like race, gender, income, and more.

The SAT, or Scholastic Aptitude Test, is an exam that U.S. high school students take before applying to college. Colleges take the test scores into account when deciding who to admit, so it's fairly important to perform well on it.

The test consists of three sections, each of which has 800 possible points. The combined score is out of 2,400 possible points (while this number has changed a few times, the data set for our project is based on 2,400 total points). **Organizations often rank high schools by their average SAT scores.** The scores are also considered a measure of overall school district quality.

New York City makes its data on [high school SAT scores](https://data.cityofnewyork.us/Education/SAT-Results/f9bf-2cp4) available online, as well as the [demographics for each high school](https://data.cityofnewyork.us/Education/DOE-High-School-Directory-2014-2015/n3p6-zve2). The first few rows of the SAT data look like this:

<left><img width="600" src="https://drive.google.com/uc?export=view&id=1_1f_s0jbjSH1YeWs-W3Ioy8mqDqaYYNb"></left>

Unfortunately, combining both of the data sets won't give us all of the demographic information we want to use. We'll need to supplement our data with other sources to do our full analysis.

The same website has several related data sets covering demographic information and test scores. Here are the links to all of the data sets we'll be using:

- [SAT scores by school](https://data.cityofnewyork.us/Education/SAT-Results/f9bf-2cp4) - SAT scores for each high school in New York City
- [School attendance](https://data.cityofnewyork.us/Education/School-Attendance-and-Enrollment-Statistics-by-Dis/7z8d-msnt) - Attendance information for each school in New York City
- [Class size](https://data.cityofnewyork.us/Education/2010-2011-Class-Size-School-level-detail/urz7-pzb3) - Information on class size for each school
- [AP test results](https://data.cityofnewyork.us/Education/AP-College-Board-2010-School-Level-Results/itfs-ms3e) - Advanced Placement (AP) exam results for each high school (passing an optional AP exam in a particular subject can earn a student college credit in that subject)
- [Graduation outcomes](https://data.cityofnewyork.us/Education/Graduation-Outcomes-Classes-Of-2005-2010-School-Le/vh2h-md7a) - The percentage of students who graduated, and other outcome information
- [Demographics](https://data.cityofnewyork.us/Education/School-Demographics-and-Accountability-Snapshot-20/ihfw-zy9j) - Demographic information for each school
- [School survey](https://data.cityofnewyork.us/Education/NYC-School-Survey-2011/mnz3-dyi8) - Surveys of parents, teachers, and students at each school
All of these data sets are interrelated. We'll need to combine them into a single data set before we can find correlations.

## 1.3 Finding Background Information

Before we move into coding, we'll need to do some background research. A thorough understanding of the data will help us avoid costly mistakes, such as thinking that a column represents something other than what it does. Background research will also give us a better understanding of how to combine and analyze the data.

In this case, we'll want to research:

- [New York City](https://en.wikipedia.org/wiki/New_York_City)
- [The SAT](https://en.wikipedia.org/wiki/SAT)
- [Schools in New York City](https://en.wikipedia.org/wiki/List_of_high_schools_in_New_York_City)
- [Our data](https://data.cityofnewyork.us/browse?category=Education)

We can learn a few different things from these resources. For example:

- Only high school students take the SAT, so we'll want to focus on high schools.
- New York City is made up of five boroughs, which are essentially distinct regions.
- New York City schools fall within several different school districts, each of which can contains dozens of schools.
- Our data sets include several different types of schools. We'll need to clean them so that we can focus on high schools only.
- Each school in New York City has a unique code called a **DBN**, or district borough number.
- Aggregating data by district will allow us to use the district mapping data to plot district-by-district differences.

## 1.4 Reading in the Data

Once we've done our background research, we're ready to read in the data. For your convenience, we've placed all the data into the schools folder. Here are all of the files in the folder:

- **ap_2010.csv** - [Data on AP test results](https://data.cityofnewyork.us/Education/AP-College-Board-2010-School-Level-Results/itfs-ms3e)
- **class_size.csv** - Data on [class size](https://data.cityofnewyork.us/Education/2010-2011-Class-Size-School-level-detail/urz7-pzb3)
- **demographics.csv** - Data on [demographics](https://data.cityofnewyork.us/Education/School-Demographics-and-Accountability-Snapshot-20/ihfw-zy9j)
- **graduation.csv** - Data on [graduation outcomes](https://data.cityofnewyork.us/Education/Graduation-Outcomes-Classes-Of-2005-2010-School-Le/vh2h-md7a)
- **hs_directory.csv** - [A directory of high schools](https://data.cityofnewyork.us/Education/DOE-High-School-Directory-2014-2015/n3p6-zve2)
- **sat_results.csv** - Data on [SAT scores](https://data.cityofnewyork.us/Education/SAT-Results/f9bf-2cp4)
- **survey_all.txt** - Data on [surveys](https://data.cityofnewyork.us/Education/NYC-School-Survey-2011/mnz3-dyi8) from all schools
- **survey_d75.txt** - Data on surveys from New York City district 75

**survey_all.txt** and **survey_d75.txt** are in more complicated formats than the other files. For now, we'll focus on reading in the - CSV files only, and then explore them.

We'll read each file into a [pandas dataframe](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html), and then store all of the dataframes in a dictionary. This will give us a convenient way to store them, and a quick way to reference them later on.

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- Read each of the files in the list **data_files** into a pandas dataframe using the **pandas.read_csv()** function.
- Recall that all of the data sets are in the **datasets** folder. That means the path to **ap_2010.csv** is **datasets/ap_2010.csv**.
- Add each of the dataframes to the **dictionary data**, using the base of **filename as the key**. For example, you'd enter **ap_2010** for the file **ap_2010.csv.**
- Afterwards, **data** should have the following keys:
    - **ap_2010**
    - **class_size**
    - **demographics**
    - **graduation**
    - **hs_directory**
    - **sat_results**
- In addition, each key in **data** should have the corresponding dataframe as its value.


In [0]:
# put your code here

import pandas as pd
data_files = [
    "ap_2010.csv",
    "class_size.csv",
    "demographics.csv",
    "graduation.csv",
    "hs_directory.csv",
    "sat_results.csv"
]

data = {}

for files in data_files:
  d = pd.read_csv("{0}".format(files))
  key_name = files.replace(".csv","")
  data[key_name] = d
  print(d.shape)




(258, 5)
(27611, 16)
(10075, 38)
(25096, 23)
(435, 64)
(478, 6)


## 1.5 Exploring the SAT Data

What we're mainly interested in is the SAT data set, which corresponds to the dictionary key **sat_results**. This data set contains the **SAT scores** for each high school in New York City. We eventually want to correlate selected information from this data set with information in the other data sets.

Let's explore **sat_results** to see what we can discover. Exploring the dataframe will help us understand the structure of the data, and make it easier for us to analyze it.

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- Display the first five rows of the SAT scores data.
    - Use the key **sat_results** to access the SAT scores dataframe stored in the dictionary **data.**
    - Use the **pandas.DataFrame.head()** method along with the **print()** function to display the first five rows of the dataframe.

In [0]:
# put your code here
data['sat_results']
print(data['sat_results'].head())                                      

      DBN                                    SCHOOL NAME  \
0  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1  01M448            UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2  01M450                     EAST SIDE COMMUNITY SCHOOL   
3  01M458                      FORSYTH SATELLITE ACADEMY   
4  01M509                        MARTA VALLE HIGH SCHOOL   

  Num of SAT Test Takers SAT Critical Reading Avg. Score SAT Math Avg. Score  \
0                     29                             355                 404   
1                     91                             383                 423   
2                     70                             377                 402   
3                      7                             414                 401   
4                     44                             390                 433   

  SAT Writing Avg. Score  
0                    363  
1                    366  
2                    370  
3                    359  
4                    38

## 1.6 Exploring the Remaining Data

When we printed the first five rows of the SAT data, the output looked like this:

```python
DBN                                    SCHOOL NAME  \
0  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES
1  01M448            UNIVERSITY NEIGHBORHOOD HIGH SCHOOL
2  01M450                     EAST SIDE COMMUNITY SCHOOL
3  01M458                      FORSYTH SATELLITE ACADEMY
4  01M509                        MARTA VALLE HIGH SCHOOL

  Num of SAT Test Takers SAT Critical Reading Avg. Score SAT Math Avg. Score  \
0                     29                             355                 404
1                     91                             383                 423
2                     70                             377                 402
3                      7                             414                 401
4                     44                             390                 433

  SAT Writing Avg. Score
0                    363
1                    366
2                    370
3                    359
4                    384
```

We can make a few observations based on this output:

- The **DBN** appears to be a unique ID for each school.
- We can tell from the first few rows of names that we only have data about high schools.
- There's only a single row for each high school, so each **DBN** is unique in the SAT data.
- We may eventually want to combine the three columns that contain SAT scores -- **SAT Critical Reading Avg.**, **Score SAT Math Avg. Score**, and **SAT Writing Avg. Score** -- into a single column to make the scores easier to analyze.
Given these observations, let's explore the other data sets to see if we can gain any insight into how to combine them.

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- Loop through each **key** in **data**. For each key:
    - Display the first five rows of the dataframe associated with the **key**.

In [0]:
# put your code here

In [0]:
for ii in data.keys():
  print(data[ii].head())

      DBN                             SchoolName AP Test Takers   \
0  01M448           UNIVERSITY NEIGHBORHOOD H.S.              39   
1  01M450                 EAST SIDE COMMUNITY HS              19   
2  01M515                    LOWER EASTSIDE PREP              24   
3  01M539         NEW EXPLORATIONS SCI,TECH,MATH             255   
4  02M296  High School of Hospitality Management               s   

  Total Exams Taken Number of Exams with scores 3 4 or 5  
0                49                                   10  
1                21                                    s  
2                26                                   24  
3               377                                  191  
4                 s                                    s  
   CSD BOROUGH SCHOOL CODE                SCHOOL NAME GRADE  PROGRAM TYPE  \
0    1       M        M015  P.S. 015 Roberto Clemente     0K       GEN ED   
1    1       M        M015  P.S. 015 Roberto Clemente     0K          CTT   
2    1

## 1.7 Reading in the Survey Data


We can make some observations based on the first few rows of each one.

- Each data set appears to either have a **DBN** column, or the information we need to create one. That means we can use a **DBN** column to combine the data sets. First we'll pinpoint matching rows from different data sets by looking for identical **DBNs**, then group all of their columns together in a single data set.
- Some fields look interesting for mapping -- particularly **Location 1**, which contains coordinates inside a larger string.
- Some of the data sets appear to contain multiple rows for each school (because the rows have duplicate **DBN** values). That means we’ll have to do some preprocessing to ensure that each **DBN** is unique within each data set. If we don't do this, we'll run into problems when we combine the data sets, because we might be merging two rows in one data set with one row in another data set.

Before we proceed with the merge, we should make sure we have all of the data we want to unify. We mentioned the survey data earlier (**survey_all.txt** and **survey_d75.txt**), but we didn't read those files in because they're in a slightly more complex format.

Each survey text file looks like this:

```python
dbn bn  schoolname  d75 studentssurveyed    highschool  schooltype  rr_s
"01M015"    "M015"  "P.S. 015 Roberto Clemente" 0   "No"    0   "Elementary School"     88
```

The files are tab delimited and encoded with **Windows-1252** encoding. An encoding defines how a computer stores the contents of a file in binary. The most common encodings are **UTF-8** and **ASCII**. **Windows-1252** is rarely used, and can cause errors if we read such a file in without specifying the encoding. If you'd like to read more about encodings, [here's](http://kunststube.net/encoding/) a good primer.

We'll need to specify the encoding and delimiter to the pandas [pandas.read_csv()](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html) function to ensure it reads the surveys in properly.

After we read in the survey data, we'll want to combine it into a single dataframe. We can do this by calling the [pandas.concat()](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html) function:

```python
z = pd.concat([x,y], axis=0)
```

The code above will combine dataframes x and y by essentially appending y to the end of x. The combined dataframe z will have the number of rows in x plus the number of rows in y.

## 1.8 Reading in the Survey Data

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- Read in **survey_all.txt.**
    - Use the **pandas.read_csv()** function to read **survey_all.txt** into the variable **all_survey**. Recall that this file is located in the **datasets** folder.
    - Specify the keyword argument **delimiter="\t".**
    - Specify the keyword argument **encoding="windows-1252".**
- Read in **survey_d75.txt.**
    - Use the **pandas.read_csv()** function to read **datasets/survey_d75.txt** into the variable **d75_survey**. Recall that this file is located in the **datasets** folder.
    - Specify the keyword argument **delimiter="\t".**
    - Specify the keyword argument **encoding="windows-1252".**
- Combine **d75_survey** and **all_survey** into a single dataframe.
    - Use the pandas [concat()](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html) function with the keyword argument **axis=0** to combine **d75_survey** and **all_survey** into the dataframe **survey**.
    - Pass in **all_survey** first, then **d75_survey** when calling the **pandas.concat()** function.
- Display the first five rows of **survey** using the **pandas.DataFrame.head()** function.



In [0]:
# put your code here
all_survey = pd.read_csv("survey_all.txt",delimiter="\t",encoding="windows-1252")
d75_survey = pd.read_csv("survey_d75.txt",delimiter="\t",encoding="windows-1252")
survey = pd.concat([all_survey,d75_survey],axis=0)
survey.head()

N_p    N_s   N_t  aca_p_11  aca_s_11  aca_t_11  aca_tot_11    bn  \
0   90.0    NaN  22.0       7.8       NaN       7.9         7.9  M015   
1  161.0    NaN  34.0       7.8       NaN       9.1         8.4  M019   
2  367.0    NaN  42.0       8.6       NaN       7.5         8.0  M020   
3  151.0  145.0  29.0       8.5       7.4       7.8         7.9  M034   
4   90.0    NaN  23.0       7.9       NaN       8.1         8.0  M063   

   com_p_11  com_s_11   ...    t_q8c_1  t_q8c_2  t_q8c_3 t_q8c_4  t_q9  \
0       7.6       NaN   ...       29.0     67.0      5.0     0.0   NaN   
1       7.6       NaN   ...       74.0     21.0      6.0     0.0   NaN   
2       8.3       NaN   ...       33.0     35.0     20.0    13.0   NaN   
3       8.2       5.9   ...       21.0     45.0     28.0     7.0   NaN   
4       7.9       NaN   ...       59.0     36.0      5.0     0.0   NaN   

   t_q9_1  t_q9_2  t_q9_3  t_q9_4  t_q9_5  
0     5.0    14.0    52.0    24.0     5.0  
1     3.0     6.0     3.0    78.0     9.0  
2     3.0     5.0    16.0    70.0     5.0  
3     0.0    18.0    32.0    39.0    11.0  
4    10.0     5.0    10.0    60.0    15.0  

[5 rows x 2773 columns]

## 1.9 Cleaning Up the Surveys

In the last step, the expected output was:

```python
    N_p  N_s  N_t  aca_p_11  aca_s_11  aca_t_11  aca_tot_11    bn  com_p_11  \
0   90  NaN   22       7.8       NaN       7.9         7.9  M015       7.6   
1  161  NaN   34       7.8       NaN       9.1         8.4  M019       7.6
```

There are two immediate facts that we can see in the data:

- There are over **2000** columns, nearly all of which we don't need. We'll have to filter the data to remove the unnecessary ones. Working with fewer columns will make it easier to print the dataframe out and find correlations within it.
- The survey data has a **dbn** column that we'll want to convert to uppercase (**DBN**). The conversion will make the column name consistent with the other data sets.

First, we'll need filter the columns to remove the ones we don't need. Luckily, there's a data dictionary at the [original data download](https://data.cityofnewyork.us/Education/NYC-School-Survey-2011/mnz3-dyi8) location. The dictionary tells us what each column represents. Based on our knowledge of the problem and the analysis we're trying to do, we can use the data dictionary to determine which columns to use.

Here's a preview of the data dictionary:


<left><img width="800" src="https://drive.google.com/uc?export=view&id=145GxwJiDOodRgyXfV15ybUBqfbDLFwEU"></left>


Based on the dictionary, it looks like these are the relevant columns:

```python
["dbn", "rr_s", "rr_t", "rr_p", "N_s", "N_t", "N_p", "saf_p_11", "com_p_11", 
 "eng_p_11", "aca_p_11", "saf_t_11", "com_t_11", "eng_t_11", 
 "aca_t_11", "saf_s_11", "com_s_11", "eng_s_11", "aca_s_11",
 "saf_tot_11", "com_tot_11", "eng_tot_11", "aca_tot_11"]
```

These columns will give us aggregate survey data about how parents, teachers, and students feel about school safety, academic performance, and more. It will also give us the **DBN**, which allows us to uniquely identify the school.

Before we filter columns out, we'll want to copy the data from the **dbn** column into a new column called **DBN**. We can copy columns like this:

```python
survey["new_column"] = survey["old_column"]
```

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- Copy the data from the **dbn** column of **survey** into a new column in **survey** called **DBN**.
- Filter **survey** so it only contains the columns we listed above. You can do this using **pandas.DataFrame.loc[]**.
    - Remember that we renamed **dbn** to **DBN**; be sure to change the list of columns we want to keep accordingly.
- Assign the dataframe **survey** to the key **survey** in the dictionary **data**.
- When you're finished, the value in **data["survey"]** should be a dataframe with 23 columns and 1702 rows.


In [0]:
# put your code here
survey["DBN"] = survey["dbn"]
survey = survey.filter(items=["DBN", "rr_s", "rr_t", "rr_p", "N_s", "N_t", "N_p", "saf_p_11", "com_p_11", 
 "eng_p_11", "aca_p_11", "saf_t_11", "com_t_11", "eng_t_11", 
 "aca_t_11", "saf_s_11", "com_s_11", "eng_s_11", "aca_s_11",
 "saf_tot_11", "com_tot_11", "eng_tot_11", "aca_tot_11"])
survey


DBN   rr_s  rr_t  rr_p    N_s    N_t    N_p  saf_p_11  com_p_11  \
0   01M015    NaN    88    60    NaN   22.0   90.0       8.5       7.6   
1   01M019    NaN   100    60    NaN   34.0  161.0       8.4       7.6   
2   01M020    NaN    88    73    NaN   42.0  367.0       8.9       8.3   
3   01M034   89.0    73    50  145.0   29.0  151.0       8.8       8.2   
4   01M063    NaN   100    60    NaN   23.0   90.0       8.7       7.9   
5   01M064    NaN    94    61    NaN   32.0  159.0       8.8       8.2   
6   01M110    NaN    91    79    NaN   32.0  302.0       8.4       7.6   
7   01M134    NaN    74    66    NaN   23.0  192.0       9.0       7.9   
8   01M137    NaN    67    75    NaN   14.0  145.0       8.7       8.0   
9   01M140   98.0    94    68  185.0   34.0  213.0       8.8       8.0   
10  01M142    NaN   100   100    NaN   38.0  331.0       9.1       8.0   
11  01M184   96.0    82    80  156.0   36.0  382.0       8.7       8.1   
12  01M188   82.0    59    40  119.0   22.0  106.0       8.7       8.1   
13  01M292   89.0    70    39  379.0   26.0  151.0       7.8       7.7   
14  01M301   99.0   100    21  181.0   22.0   38.0       8.5       7.3   
15  01M315    NaN   100    67    NaN   20.0  135.0       9.0       8.6   
16  01M332   96.0   100    86  110.0   14.0   93.0       8.4       7.9   
17  01M345  100.0    87    53  254.0   20.0  127.0       8.4       7.6   
18  01M361    NaN    74    42    NaN   17.0   93.0       8.9       8.2   
19  01M363    NaN    54    40    NaN   13.0  115.0       8.7       8.1   
20  01M364    NaN    86    54    NaN   19.0  152.0       9.3       8.9   
21  01M378   78.0    91    82  204.0   20.0  203.0       8.8       7.9   
22  01M448   84.0    95    10  385.0   37.0   46.0       7.9       7.4   
23  01M450    0.0    98    28    NaN   42.0  150.0       8.7       8.2   
24  01M458   40.0   100    23   66.0   10.0   37.0       8.1       7.0   
25  01M509   90.0   100    21  306.0   29.0   69.0       7.7       7.4   
26  01M515   95.0   100    86  454.0   36.0  389.0       8.3       7.2   
27  01M539   98.0    68    51  923.0   67.0  736.0       8.5       7.9   
28  01M650   75.0    86    19  125.0   12.0   30.0       9.0       8.4   
29  01M696   98.0    81    50  561.0   30.0  283.0       8.8       8.2   
..     ...    ...   ...   ...    ...    ...    ...       ...       ...   
26  75M811   62.0    61    54   44.0   54.0  172.0       8.8       8.4   
27  75Q004   89.0    92    54   24.0   95.0  198.0       9.0       8.4   
28  75Q009   57.0    61    39   51.0   59.0  163.0       8.9       8.2   
29  75Q023   75.0    93    21   93.0   65.0   52.0       8.7       8.0   
30  75Q075   40.0    69    28   23.0   57.0   85.0       8.7       7.7   
31  75Q177   64.0    70    41   95.0   85.0  197.0       8.9       8.0   
32  75Q224   97.0    75    49   62.0   75.0  205.0       9.0       8.3   
33  75Q233   85.0    73    26  121.0   56.0  101.0       8.9       8.0   
34  75Q255   71.0    60    48   12.0   76.0  194.0       8.9       8.2   
35  75Q256   88.0    93    46  195.0  105.0  184.0       8.6       8.0   
36  75Q721   90.0    77    43  195.0   61.0  172.0       8.9       8.1   
37  75Q752   45.0    47     9  160.0   36.0   34.0       8.5       7.5   
38  75Q811   75.0    71    34  101.0   49.0  111.0       8.8       8.0   
39  75Q993   87.0    76    47   90.0   71.0  185.0       9.2       8.5   
40  75R025   57.0    75    21  238.0   70.0   97.0       8.1       7.5   
41  75R037    0.0    55    46    NaN   52.0  143.0       8.9       8.0   
42  75R373   92.0    60    34   22.0   64.0  132.0       8.8       8.1   
43  75R721   96.0    93    57   82.0   71.0  187.0       9.1       8.5   
44  75X010   84.0    91    56   76.0   86.0  247.0       9.0       8.5   
45  75X012   19.0    47     9   37.0   25.0   18.0       5.8       6.5   
46  75X017   69.0    79    37   64.0   72.0  122.0       8.8       8.1   
47  75X168   92.0    71    53   78.0   61.0  206.0       8.9       8.5   
48  75X176 

## 1.10 Inserting DBN Fields

When we explored all of the data sets, we noticed that some of them, like **class_size** and **hs_directory**, don't have a DBN column. **hs_directory** does have a **dbn** column, though, so we can just rename it.

However, **class_size** doesn't appear to have the column at all. Here are the first few rows of the data set:

```python
    CSD BOROUGH   SCHOOL CODE         SCHOOL NAME GRADE  PROGRAM TYPE  \
0    1       M        M015  P.S. 015 Roberto Clemente     0K       GEN ED
1    1       M        M015  P.S. 015 Roberto Clemente     0K          CTT
2    1       M        M015  P.S. 015 Roberto Clemente     01       GEN ED
3    1       M        M015  P.S. 015 Roberto Clemente     01          CTT
4    1       M        M015  P.S. 015 Roberto Clemente     02       GEN ED
```

Here are the first few rows of the **sat_results** data, which does have a **DBN** column:

```python
DBN                                    SCHOOL NAME  \
0  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES
1  01M448            UNIVERSITY NEIGHBORHOOD HIGH SCHOOL
2  01M450                     EAST SIDE COMMUNITY SCHOOL
3  01M458                      FORSYTH SATELLITE ACADEMY
4  01M509                        MARTA VALLE HIGH SCHOOL
```

From looking at these rows, we can tell that the **DBN** in the **sat_results** data is just a combination of the **CSD** and **SCHOOL CODE** columns in the **class_size** data. The main difference is that the **DBN** is padded, so that the **CSD** portion of it always consists of two digits. That means we'll need to add a leading 0 to the **CSD** if the **CSD** is less than two digits long. Here's a diagram illustrating what we need to do:

<center><img width="150" src="https://drive.google.com/uc?export=view&id=1k1QuSctrJnMW6gfAbTs-DJbzTe23en2U"></center>

As you can see, whenever the **CSD** is less than two digits long, we need to add a leading 0. We can accomplish this using the **pandas.Series.apply()** method, along with a custom function that:

- Takes in a number.
- Converts the number to a string using the **str()** function.
- Check the length of the string using the **len()** function.
    - If the string is two digits long, returns the string.
    - If the string is one digit long, adds a 0 to the front of the string, then returns it.
        - You can use the string method [zfill()](https://docs.python.org/3/library/stdtypes.html#str.zfill) to do this.

Once we've padded the **CSD**, we can use the addition operator (+) to combine the values in the **CSD** and **SCHOOL CODE** columns. Here's an example of how we would do this:

```python
dataframe["new_column"] = dataframe["column_one"] + dataframe["column_two"]
```

And here's a diagram illustrating the basic concept:

<center><img width="300" src="https://drive.google.com/uc?export=view&id=1Io7o-45Pixlv2tX8rTyZvL6KjIq9qMPz"></center>


## 1.11 Inserting DBN Fields

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- Copy the **dbn** column in **hs_directory** into a new column called **DBN**.
- Create a new column called **padded_csd** in the **class_size** data set.
    - Use the [pandas.Series.apply()](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.apply.html) method along with the function **pad_csd()** to generate this column.
        - Make sure to apply the function along the **data["class_size"]["CSD"]** column.
- Use the addition operator (+) along with the **padded_csd** and **SCHOOL CODE** columns of **class_size**, then assign the result to the **DBN** column of **class_size**.
- Display the first few rows of **class_size** to double check the **DBN** column.

In [0]:
# put your code here
def pad_csd(num):
    string_representation = str(num)
    if len(string_representation) > 1:
        return string_representation
    else:
        return string_representation.zfill(2)

data['hs_directory']['DBN'] = data['hs_directory']['dbn']
data["class_size"]["padded_csd"] = data["class_size"]["CSD"].apply(pad_csd)
data['class_size']['DBN'] = data['class_size']['padded_csd'] + data['class_size']['SCHOOL CODE']
data['class_size']


CSD BOROUGH SCHOOL CODE  \
0        1       M        M015   
1        1       M        M015   
2        1       M        M015   
3        1       M        M015   
4        1       M        M015   
5        1       M        M015   
6        1       M        M015   
7        1       M        M015   
8        1       M        M015   
9        1       M        M015   
10       1       M        M015   
11       1       M        M015   
12       1       M        M019   
13       1       M        M019   
14       1       M        M019   
15       1       M        M019   
16       1       M        M019   
17       1       M        M019   
18       1       M        M019   
19       1       M        M019   
20       1       M        M019   
21       1       M        M019   
22       1       M        M019   
23       1       M        M019   
24       1       M        M019   
25       1       M        M020   
26       1       M        M020   
27       1       M        M020   
28       1       M        M020   
29       1       M        M020   
...    ...     ...         ...   
27581   32       K        K556   
27582   32       K        K556   
27583   32       K        K556   
27584   32       K        K556   
27585   32       K        K556   
27586   32       K        K556   
27587   32       K        K556   
27588   32       K        K556   
27589   32       K        K556   
27590   32       K        K556   
27591   32       K        K556   
27592   32       K        K556   
27593   32       K        K556   
27594   32       K        K556   
27595   32       K        K564   
27596   32       K        K564   
27597   32       K        K564   
27598   32       K        K564   
27599   32       K        K564   
27600   32       K        K564   
27601   32       K        K564   
27602   32       K        K564   
27603   32       K        K564   
27604   32       K        K564   
27605   32       K        K564   
27606   32       K        K564   
27607   32       K        K564   
27608   32       K        K564   
27609   32       K        K564   
27610   32       K        K564   

                                             SCHOOL NAME GRADE  PROGRAM TYPE  \
0                              P.S. 015 Roberto Clemente     0K       GEN ED   
1                              P.S. 015 Roberto Clemente     0K          CTT   
2                              P.S. 015 Roberto Clemente     01       GEN ED   
3                              P.S. 015 Roberto Clemente     01          CTT   
4                              P.S. 015 Roberto Clemente     02       GEN ED   
5                              P.S. 015 Roberto Clemente     02          CTT   
6                              P.S. 015 Roberto Clemente     03       GEN ED   
7                              P.S. 015 Roberto Clemente     03          CTT   
8                              P.S. 015 Roberto Clemente     04       GEN ED   
9                              P.S. 015 Roberto Clemente     05       GEN ED   
10                             P.S. 015 Roberto Clemente  0K-09      SPEC ED   
11                             P.S. 015 Roberto Clemente    NaN          NaN   
12                                   P.S. 019 Asher Levy     0K       GEN ED   
13                                   P.S. 019 Asher Levy     0K          CTT   
14                                   P.S. 019 Asher Levy     01       GEN ED   
15                                   P.S. 019 Asher Levy     01          CTT   
16                                   P.S. 019 Asher Levy     02       GEN ED   
17                                   P.S. 019 Asher Levy     03       GEN ED   
18                                   P.S. 019 Asher Levy     04       GEN ED   
19                                   P.S. 019 Asher Levy     04          CTT   
20                                   P.S. 019 Asher Levy     05       GEN ED   
21                                   P.S. 019 Asher Levy     05          CTT   
22                                   P.S. 019 Asher Levy  

## 1.12 Combining the SAT Scores


Now we're almost ready to combine our data sets. Before we do, let's take some time to calculate variables that will be useful in our analysis. We've already discussed one such variable -- a column that totals up the SAT scores for the different sections of the exam. This will make it much easier to correlate scores with demographic factors because we'll be working with a single number, rather than three different ones.

Before we can generate this column, we'll need to convert the **SAT Math Avg. Score**, **SAT Critical Reading Avg. Score**, and **SAT Writing Avg. Score** columns in the **sat_results** data set from the object (string) data type to a numeric data type. We can use the [pandas.to_numeric()](http://pandas.pydata.org/pandas-docs/version/0.17.0/generated/pandas.to_numeric.html) method for the conversion. If we don't convert the values, we won't be able to add the columns together.

It's important to pass the keyword argument **errors="coerce"** when we call **pandas.to_numeric()**, so that pandas treats any invalid strings it can't convert to numbers as missing values instead.

After we perform the conversion, we can use the addition operator (+) to add all three columns together.


**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- Convert the **SAT Math Avg. Score**, **SAT Critical Reading Avg. Score**, and **SAT Writing Avg. Score** columns in the **sat_results** data set from the object (string) data type to a numeric data type.
    - Use the **pandas.to_numeric()** function on each of the columns, and assign the result back to the same column.
    - Pass in the keyword argument **errors="coerce".**
- Create a column called **sat_score** in **sat_results** that holds the combined SAT score for each student.
    - Add up **SAT Math Avg. Score**, **SAT Critical Reading Avg. Score**, and **SAT Writing Avg. Score**, and assign the total to the **sat_score** column of **sat_results.**
- Display the first few rows of the **sat_score column** of **sat_results** to verify that everything went okay.

In [0]:
# put your code here
data["sat_results"]['SAT Math Avg. Score'] = pd.to_numeric(data["sat_results"]['SAT Math Avg. Score'],errors="coerce")
data["sat_results"]['SAT Critical Reading Avg. Score'] = pd.to_numeric(data["sat_results"]['SAT Critical Reading Avg. Score'],errors="coerce")
data["sat_results"]['SAT Writing Avg. Score'] = pd.to_numeric(data["sat_results"]['SAT Writing Avg. Score'],errors="coerce")
data["sat_results"]["sat_score"] = data["sat_results"]['SAT Math Avg. Score'] + data["sat_results"]['SAT Critical Reading Avg. Score'] + data["sat_results"]['SAT Writing Avg. Score']
data["sat_results"]

DBN                                        SCHOOL NAME  \
0    01M292      HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1    01M448                UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2    01M450                         EAST SIDE COMMUNITY SCHOOL   
3    01M458                          FORSYTH SATELLITE ACADEMY   
4    01M509                            MARTA VALLE HIGH SCHOOL   
5    01M515            LOWER EAST SIDE PREPARATORY HIGH SCHOOL   
6    01M539  NEW EXPLORATIONS INTO SCIENCE, TECHNOLOGY AND ...   
7    01M650                               CASCADES HIGH SCHOOL   
8    01M696                     BARD HIGH SCHOOL EARLY COLLEGE   
9    02M047  47 THE AMERICAN SIGN LANGUAGE AND ENGLISH SECO...   
10   02M288                       FOOD AND FINANCE HIGH SCHOOL   
11   02M294                               ESSEX STREET ACADEMY   
12   02M296              HIGH SCHOOL OF HOSPITALITY MANAGEMENT   
13   02M298                                   PACE HIGH SCHOOL   
14   02M300  URBAN ASSEMBLY SCHOOL OF DESIGN AND CONSTRUCTI...   
15   02M303                         FACING HISTORY SCHOOL, THE   
16   02M305  URBAN ASSEMBLY ACADEMY OF GOVERNMENT AND LAW, THE   
17   02M308                       LOWER MANHATTAN ARTS ACADEMY   
18   02M313  JAMES BALDWIN SCHOOL, THE: A SCHOOL FOR EXPEDI...   
19   02M316  URBAN ASSEMBLY SCHOOL OF BUSINESS FOR YOUNG WO...   
20   02M374                          GRAMERCY ARTS HIGH SCHOOL   
21   02M376                                        NYC ISCHOOL   
22   02M392                         MANHATTAN BUSINESS ACADEMY   
23   02M393                          BUSINESS OF SPORTS SCHOOL   
24   02M394                           EMMA LAZARUS HIGH SCHOOL   
25   02M399         THE HIGH SCHOOL FOR LANGUAGE AND DIPLOMACY   
26   02M400              HIGH SCHOOL FOR ENVIRONMENTAL STUDIES   
27   02M407              INSTITUTE FOR COLLABORATIVE EDUCATION   
28   02M408           PROFESSIONAL PERFORMING ARTS HIGH SCHOOL   
29   02M411                  BARUCH COLLEGE CAMPUS HIGH SCHOOL   
..      ...                                                ...   
448  30Q580          BACCALAUREATE SCHOOL FOR GLOBAL EDUCATION   
449  31R047          CSI HIGH SCHOOL FOR INTERNATIONAL STUDIES   
450  31R064        GAYNOR MCCOWN EXPEDITIONARY LEARNING SCHOOL   
451  31R080                     THE MICHAEL J. PETRIDES SCHOOL   
452  31R440                               NEW DORP HIGH SCHOOL   
453  31R445                          PORT RICHMOND HIGH SCHOOL   
454  31R450                                 CURTIS HIGH SCHOOL   
455  31R455                            TOTTENVILLE HIGH SCHOOL   
456  31R460                        SUSAN E. WAGNER HIGH SCHOOL   
457  31R470                                CONCORD HIGH SCHOOL   
458  31R600  RALPH R. MCKEE CAREER AND TECHNICAL EDUCATION ...   
459  31R605                STATEN ISLAND TECHNICAL HIGH SCHOOL   
460  31R607    YOUNG ADULT BOROUGH CENTER AT STATEN IS TECH HS   
461  32K403               ACADEMY FOR ENVIRONMENTAL LEADERSHIP   
462  32K545        EBC HIGH SCHOOL FOR PUBLIC SERVICE–BUSHWICK   
463  32K549                 BUSHWICK SCHOOL FOR SOCIAL JUSTICE   
464  32K552                          ACADEMY OF URBAN PLANNING   
465  32K554               ALL CITY LEADERSHIP SECONDARY SCHOOL   
466  32K556  BUSHWICK LEADERS HIGH SCHOOL FOR ACADEMIC EXCE...   
467  32K564                     BUSHWICK COMMUNITY HIGH SCHOOL   
468  75K371                      P.S. 371 s LILLIAN L. RASHKIS   
469  75M035                                           P.S. 035   
470  75Q256                                          P.S. Q256   
471  75Q811                                          P.S. Q811   
472  75R025            SOUTH RICHMOND HIGH SCHOOL I.S./P.S. 25   
473  75X012                   P.S. X012 LEWIS AND CLARK SCHOOL   
474  75X754            J. M. RAPPORT SCHOOL CAREER DEVELOPMENT   
475  79M645         SCHOOL FOR COOPERATIVE TECHNICAL EDUCATION   
476  79Q950                                GED P

## 1.13 Parsing Geographic Coordinates for Schools

Next, we'll want to parse the **latitude** and **longitude** coordinates for each school. This will enable us to map the schools and uncover any geographic patterns in the data. The coordinates are currently in the text field **Location 1** in the **hs_directory** data set.

Let's take a look at the first few rows:

```python
0    883 Classon Avenue\nBrooklyn, NY 11225\n(40.67...
1    1110 Boston Road\nBronx, NY 10456\n(40.8276026...
2    1501 Jerome Avenue\nBronx, NY 10452\n(40.84241...
3    411 Pearl Street\nNew York, NY 10038\n(40.7106...
4    160-20 Goethals Avenue\nJamaica, NY 11432\n(40...
```        
                                               
As you can see, this field contains a lot of information we don't need. We want to extract the coordinates, which are in parentheses at the end of the field. Here's an example:

```python
1110 Boston Road\nBronx, NY 10456\n(40.8276026690005, -73.90447525699966)
```

We want to extract the **latitude**, 40.8276026690005, and the **longitude**, -73.90447525699966. Taken together, **latitude** and **longitude** make up a pair of coordinates that allows us to pinpoint any location on Earth.

We can do the extraction with a **regular expression**. The following expression will pull out everything inside the parentheses:

```python
import re
re.findall("\(.+\)", "1110 Boston Road\nBronx, NY 10456\n(40.8276026690005, -73.90447525699966)")
```

This command will return (40.8276026690005, -73.90447525699966). We'll need to process this result further using the string methods [split()](https://docs.python.org/3/library/stdtypes.html#str.split) and [replace()](https://docs.python.org/3/library/stdtypes.html#str.replace) methods to extract each coordinate.

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- Write a function that:
    - Takes in a string
    - Uses the regular expression above to extract the coordinates
    - Uses string manipulation functions to pull out the latitude
    - Returns the latitude
- Use the [Series.apply()](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.apply.html) method to apply the function across the **Location 1** column of **hs_directory**. Assign the result to the **lat** column of **hs_directory.**
- Display the first few rows of **hs_directory** to verify the results.

In [0]:
# put your code here
import re
def school_latitude(lat):
  lat = re.findall("\(.+\)", lat)
  lat = re.findall("\((.*?)\,",lat[0])
  return lat[0]

data["hs_directory"]["lat"] = data['hs_directory']['Location 1'].apply(school_latitude)
data['hs_directory']




dbn                                        school_name        borough  \
0    17K548                Brooklyn School for Music & Theatre       Brooklyn   
1    09X543                   High School for Violin and Dance          Bronx   
2    09X327        Comprehensive Model School Project M.S. 327          Bronx   
3    02M280     Manhattan Early College School for Advertising      Manhattan   
4    28Q680  Queens Gateway to Health Sciences Secondary Sc...         Queens   
5    08X348              Schuylerville Preparatory High School          Bronx   
6    14K474      PROGRESS High School for Professional Careers       Brooklyn   
7    02M420  High School for Health Professions and Human S...      Manhattan   
8    28Q350                    Jamaica Gateway to the Sciences         Queens   
9    29Q313  Benjamin Franklin High School for Finance & In...         Queens   
10   12X271                  East Bronx Academy for the Future          Bronx   
11   02M520   Murry Bergtraum High School for Business Careers      Manhattan   
12   06M211  Inwood Early College for Health and Informatio...      Manhattan   
13   24Q600        Queens Vocational and Technical High School         Queens   
14   10X368           In-Tech Academy (M.S. / High School 368)          Bronx   
15   06M348   Washington Heights Expeditionary Learning School      Manhattan   
16   18K629         Cultural Academy for the Arts and Sciences       Brooklyn   
17   03M299      High School for Arts, Imagination and Inquiry      Manhattan   
18   07X520         Foreign Language Academy of Global Studies          Bronx   
19   08X312                             Millennium Art Academy          Bronx   
20   08X293  Renaissance High School for Musical Theater & ...          Bronx   
21   13K419  Science Skills Center High School for Science,...       Brooklyn   
22   31R460                        Susan E. Wagner High School  Staten Island   
23   28Q328               High School for Community Leadership         Queens   
24   05M362                          Columbia Secondary School      Manhattan   
25   10X243                  West Bronx Academy for the Future          Bronx   
26   03M307       Urban Assembly School for Media Studies, The      Manhattan   
27   20K609         Urban Assembly School for Criminal Justice       Brooklyn   
28   28Q284                         York Early College Academy         Queens   
29   21K348                   High School of Sports Management       Brooklyn   
..      ...                                                ...            ...   
405  24Q264                  Academy of Finance and Enterprise         Queens   
406  08X519  Felisa Rincon de Gautier Institute for Law and...          Bronx   
407  02M374                          Gramercy Arts High School      Manhattan   
408  17K533                School for Democracy and Leadership       Brooklyn   
409  26Q415                    Benjamin N. Cardozo High School         Queens   
410  31R455                            Tottenville High School  Staten Island   
411  25Q263                 Flushing International High School         Queens   
412  02M449                               Vanguard High School      Manhattan   
413  32K552                          Academy of Urban Planning       Brooklyn   
414  06M468        High School for Health Careers and Sciences      Manhattan   
415  30Q580          Baccalaureate School for Global Education         Queens   
416  12X684                                      Wings Academy          Bronx   
417  22K405                                Midwood High School       Brooklyn   
418  25Q670            Robert F. Kennedy Community High School         Queens   
419  19K618                   Academy of Innovative Technology       Brooklyn   
420  02M298                                   Pace High School      Manhattan   
421  08X376  Antonia Pantoja Preparatory Academy: A College...          Bronx   
422  02M500                Unity Center

In [0]:
!ls

'ap_2010 (1).csv'	'graduation (1).csv'	 sat_results.csv
 ap_2010.csv		 graduation.csv		'survey_all (1).txt'
'class_size (1).csv'	'hs_directory (1).csv'	 survey_all.txt
 class_size.csv		 hs_directory.csv	'survey_d75 (1).txt'
'demographics (1).csv'	 sample_data		 survey_d75.txt
 demographics.csv	'sat_results (1).csv'


## 1.14 Extracting the Longitude

On the last screen, we parsed the **latitude** from the **Location 1** column. Now we'll just need to do the same for the longitude.

Once we have both coordinates, we'll need to convert them to numeric values. We can use the **pandas.to_numeric()** function to convert them from strings to numbers.

**Exercise**

<left><img width="100" src="https://drive.google.com/uc?export=view&id=1E8tR7B9YYUXsU_rddJAyq0FrM0MSelxZ"></left>

- Write a function that:
    - Takes in a string.
    - Uses the regular expression above to extract the coordinates.
    - Uses string manipulation functions to pull out the longitude.
    - Returns the longitude.
- Use the **Series.apply()** method to apply the function across the **Location 1** column of **hs_directory**. Assign the result to the **lon** column of **hs_directory.**
- Use the **to_numeric()** function to convert the **lat** and **lon** columns of **hs_directory** to numbers.
    - Specify the **errors="coerce"** keyword argument to handle missing values properly.
- Display the first few rows of hs_directory to verify the results.

In [0]:
# put your code here
def school_longitude(lon):
  lon = re.findall("\(.+\)", lon)
  lon = re.findall("\,(.*?)\)",lon[0])
  return lon[0]

data["hs_directory"]["lon"] = data['hs_directory']['Location 1'].apply(school_longitude)
data["hs_directory"]['lon'] = pd.to_numeric(data["hs_directory"]['lon'],errors="coerce")
data["hs_directory"]['lat'] = pd.to_numeric(data["hs_directory"]['lat'],errors="coerce")

data['hs_directory']



dbn                                        school_name        borough  \
0    17K548                Brooklyn School for Music & Theatre       Brooklyn   
1    09X543                   High School for Violin and Dance          Bronx   
2    09X327        Comprehensive Model School Project M.S. 327          Bronx   
3    02M280     Manhattan Early College School for Advertising      Manhattan   
4    28Q680  Queens Gateway to Health Sciences Secondary Sc...         Queens   
5    08X348              Schuylerville Preparatory High School          Bronx   
6    14K474      PROGRESS High School for Professional Careers       Brooklyn   
7    02M420  High School for Health Professions and Human S...      Manhattan   
8    28Q350                    Jamaica Gateway to the Sciences         Queens   
9    29Q313  Benjamin Franklin High School for Finance & In...         Queens   
10   12X271                  East Bronx Academy for the Future          Bronx   
11   02M520   Murry Bergtraum High School for Business Careers      Manhattan   
12   06M211  Inwood Early College for Health and Informatio...      Manhattan   
13   24Q600        Queens Vocational and Technical High School         Queens   
14   10X368           In-Tech Academy (M.S. / High School 368)          Bronx   
15   06M348   Washington Heights Expeditionary Learning School      Manhattan   
16   18K629         Cultural Academy for the Arts and Sciences       Brooklyn   
17   03M299      High School for Arts, Imagination and Inquiry      Manhattan   
18   07X520         Foreign Language Academy of Global Studies          Bronx   
19   08X312                             Millennium Art Academy          Bronx   
20   08X293  Renaissance High School for Musical Theater & ...          Bronx   
21   13K419  Science Skills Center High School for Science,...       Brooklyn   
22   31R460                        Susan E. Wagner High School  Staten Island   
23   28Q328               High School for Community Leadership         Queens   
24   05M362                          Columbia Secondary School      Manhattan   
25   10X243                  West Bronx Academy for the Future          Bronx   
26   03M307       Urban Assembly School for Media Studies, The      Manhattan   
27   20K609         Urban Assembly School for Criminal Justice       Brooklyn   
28   28Q284                         York Early College Academy         Queens   
29   21K348                   High School of Sports Management       Brooklyn   
..      ...                                                ...            ...   
405  24Q264                  Academy of Finance and Enterprise         Queens   
406  08X519  Felisa Rincon de Gautier Institute for Law and...          Bronx   
407  02M374                          Gramercy Arts High School      Manhattan   
408  17K533                School for Democracy and Leadership       Brooklyn   
409  26Q415                    Benjamin N. Cardozo High School         Queens   
410  31R455                            Tottenville High School  Staten Island   
411  25Q263                 Flushing International High School         Queens   
412  02M449                               Vanguard High School      Manhattan   
413  32K552                          Academy of Urban Planning       Brooklyn   
414  06M468        High School for Health Careers and Sciences      Manhattan   
415  30Q580          Baccalaureate School for Global Education         Queens   
416  12X684                                      Wings Academy          Bronx   
417  22K405                                Midwood High School       Brooklyn   
418  25Q670            Robert F. Kennedy Community High School         Queens   
419  19K618                   Academy of Innovative Technology       Brooklyn   
420  02M298                                   Pace High School      Manhattan   
421  08X376  Antonia Pantoja Preparatory Academy: A College...          Bronx   
422  02M500                Unity Center

In [0]:
data

{'ap_2010':         DBN                                         SchoolName  \
 0    01M448                       UNIVERSITY NEIGHBORHOOD H.S.   
 1    01M450                             EAST SIDE COMMUNITY HS   
 2    01M515                                LOWER EASTSIDE PREP   
 3    01M539                     NEW EXPLORATIONS SCI,TECH,MATH   
 4    02M296              High School of Hospitality Management   
 5    02M298                                   Pace High School   
 6    02M300  Urban Assembly School of Design and Construction,   
 7    02M303                         Facing History School, The   
 8    02M305  Urban Assembly Academy of Government and Law, The   
 9    02M308                       Lower Manhattan Arts Academy   
 10   02M400                       HS FOR ENVIRONMENTAL STUDIES   
 11   02M408                       PROFESSIONAL PERFORMING ARTS   
 12   02M411                           BARUCH COLLEGE CAMPUS HS   
 13   02M412                       NYC LAB HS FOR C

## 1.15 Next Steps

We're almost ready to combine our data sets! We've come a long way in this mission -- we've gone from choosing a topic for a project to acquiring the data to having clean data that we're almost ready to combine.

Along the way, we've learned how to:

- Handle files with different formats and columns
- Prepare to merge multiple files
- Use text processing to extract coordinates from a string
- Convert columns from strings to numbers

You'll always learn something new while working on a real-world data science project. Each project is unique, and there will always be quirks you don't quite know how to handle. The key is to be willing to try different approaches, and to have a general framework in your head for how to move from Step A to Step B.

In the next section,  we'll finish cleaning the data sets, then combine them so we can start our analysis.